File: Etapa 2/sh1106.py
from machine import Pin, I2C
import framebuf

class SH1106_I2C(framebuf.FrameBuffer):
    def __init__(self, width, height, i2c, addr=0x3C, external_vcc=False):
        self.width = width
        self.height = height
        self.i2c = i2c
        self.addr = addr
        self.external_vcc = external_vcc
        self.pages = self.height // 8
        self.buffer = bytearray(self.pages * self.width)
        super().__init__(self.buffer, self.width, self.height, framebuf.MONO_VLSB)
        self.init_display()

    def write_cmd(self, cmd):
        self.i2c.writeto(self.addr, bytearray([0x00, cmd]))

    def write_data(self, buf):
        self.i2c.writeto(self.addr, b'\x40' + buf)

    def init_display(self):
        for cmd in (
            0xAE,  # display off
            0xA8, self.height - 1,
            0xD3, 0x00,
            0x40,  # start line
            0xA1,  # segment remap
            0xC8,  # COM scan direction
            0xDA, 0x12,
            0x81, 0xCF,
            0xD9, 0xF1,
            0xDB, 0x40,
            0xA4,  # display follows RAM
            0xA6,  # normal display
            0xAF,  # display on
        ):
            self.write_cmd(cmd)
        self.fill(0)
        self.show()

    def show(self):
        for page in range(0, self.pages):
            self.write_cmd(0xB0 + page)
            self.write_cmd(0x02)      # lower column start (shift)
            self.write_cmd(0x10)      # higher column start
            start = self.width * page
            end = start + self.width
            self.write_data(self.buffer[start:end])

File: Etapa 2/Codigo C/sh1106.c
#include <stdio.h>
#include "pico/stdlib.h"
#include "hardware/i2c.h"
#include "sh1106.h"
#include "font5x7.inc"

// Buffer de pantalla
uint8_t sh1106_buffer[SH1106_BUFFERSIZE];

// ---- Funciones internas ----
static void sh1106_send_cmd(uint8_t cmd) {
    uint8_t data[2] = {0x00, cmd};
    i2c_write_blocking(SH1106_I2C_PORT, SH1106_I2C_ADDR, data, 2, false);
}

static void sh1106_send_data(uint8_t data) {
    uint8_t buf[2] = {0x40, data};
    i2c_write_blocking(SH1106_I2C_PORT, SH1106_I2C_ADDR, buf, 2, false);
}

// ---- Inicializar OLED ----
void sh1106_init() {
    sleep_ms(50);

    sh1106_send_cmd(0xAE);
    sh1106_send_cmd(0xD5); sh1106_send_cmd(0x80);
    sh1106_send_cmd(0xA8); sh1106_send_cmd(0x3F);
    sh1106_send_cmd(0xD3); sh1106_send_cmd(0x00);
    sh1106_send_cmd(0x40);
    sh1106_send_cmd(0xAD); sh1106_send_cmd(0x8B);
    sh1106_send_cmd(0xA1);
    sh1106_send_cmd(0xC8);
    sh1106_send_cmd(0xDA); sh1106_send_cmd(0x12);
    sh1106_send_cmd(0x81); sh1106_send_cmd(0x7F);
    sh1106_send_cmd(0xD9); sh1106_send_cmd(0xF1);
    sh1106_send_cmd(0xDB); sh1106_send_cmd(0x40);
    sh1106_send_cmd(0xA4);
    sh1106_send_cmd(0xA6);
    sh1106_send_cmd(0xAF);

    sh1106_clear();
    sh1106_update();
}

// ---- Limpiar pantalla ----
void sh1106_clear() {
    for (int i = 0; i < SH1106_BUFFERSIZE; i++)
        sh1106_buffer[i] = 0x00;
}

// ---- Actualizar OLED ----
void sh1106_update() {
    for (uint8_t page = 0; page < 8; page++) {
        sh1106_send_cmd(0xB0 + page);
        sh1106_send_cmd(0x02);
        sh1106_send_cmd(0x10);

        for (uint8_t col = 0; col < 128; col++) {
            sh1106_send_data(sh1106_buffer[page * 128 + col]);
        }
    }
}

// ---- Dibujar un carácter ----
void sh1106_draw_char(int x, int y, char c) {
    if (c < 32 || c > 126) return;

    const uint8_t *glyph = font5x7[c - 32];

    for (int col = 0; col < 5; col++) {
        uint8_t column = glyph[col];
        int index = x + col + (y / 8) * 128;
        if (index >= 0 && index < SH1106_BUFFERSIZE)
            sh1106_buffer[index] = column;
    }
}

// ---- Dibujar texto ----
void sh1106_draw_text(int x, int y, const char *txt) {
    while (*txt) {
        sh1106_draw_char(x, y, *txt);
        x += 6;
        txt++;
    }
}
